src/commands/crashes/upload-mappings.ts

```diff
@@ -0,0 +1,88 @@
+import { inspect } from "util";
+import { DefaultApp } from "../../util/profile";
+import UploadSymbolsHelper from "./lib/symbols-uploading-helper";
+import { SymbolType } from "./lib/symbols-uploading-helper";
+
+import * as Fs from "fs";
+import * as Path from "path";
+
+import * as _ from "lodash";
+
+const debug = require("debug")("appcenter-cli:commands:apps:crashes:upload-mappings");
+
+@help("Upload the Android mappings for the application")
+export default class UploadMappings extends AppCommand {
+  @help("Path to an Android mapping.txt file.")
+  @shortName("m")
+  @longName("mapping")
+  @hasArg
+  public mappingPath: string;
+
+  @help("The version name to associate with the mappings.")
+  @shortName("n")
+  @longName("version-name")
+  @hasArg
+  public versionName: string;
+
+  @help("The version code to associate with the mappings.")
+  @shortName("c")
+  @longName("version-code")
+  @hasArg
+  public versionCode: string;
+
+  public async run(client: AppCenterClient): Promise<CommandResult> {
+    const app: DefaultApp = this.app;
+
+    this.validateParameters();
+
+    this.getStatsForFsPath(this.mappingPath);
+    if (this.getLowerCasedFileExtension(this.mappingPath) !== ".txt") {
+      throw failure(ErrorCodes.InvalidParameter, `path ${this.mappingPath} does not point to valid mapping.txt file`);
+    }
+    const uploadRequest: models.SymbolUploadBeginRequest = {
+      symbolType: SymbolType.AndroidProGuard,
+      fileName: "mapping.txt",
+      version: this.versionName,
+      build: this.versionCode
+    };
+
+    // upload mappings
+    await out.progress("Uploading mappings...", new UploadSymbolsHelper(client, app, debug).uploadSymbolsArtifact(this.mappingPath, uploadRequest));
+
+    return success();
+  }
+
+  private getStatsForFsPath(filePath: string): Fs.Stats | null {
+    // take fs entry stats (and check it's existence BTW)
+    try {
+      debug(`Getting FS statistics for ${filePath}`);
+      return Fs.statSync(filePath);
+    } catch (error) {
+      if (error.code === "ENOENT") {
+        // path points to non-existing file system entry
+        throw failure(ErrorCodes.InvalidParameter, `path ${filePath} points to non-existent item`);
+      } else {
+        // other errors
+        debug(`Failed to get statistics for file system entry ${filePath} - ${inspect(error)}`);
+        throw failure(ErrorCodes.Exception, `failed to get statistics for file system entry ${filePath}`);
+      }
+    }
+  }
+
+  private getLowerCasedFileExtension(filePath: string): string {
+    return Path.extname(filePath).toLowerCase();
+  }
+
+  private validateParameters() {
+    // check that user have selected all of --mapping, --version-name, and --version-code
+    if (_.isNil(this.mappingPath) || _.isNil(this.versionName) || _.isNil(this.versionCode)) {
+      throw failure(ErrorCodes.InvalidParameter, "all of '--mapping', '--version-name', and '--version-code' are required");
```

Comment:
Just a thought that I haven't had a chance to dive further into: do we typically group error messages together in the CLI? Or do we fail on each specific parameter that is missing?

Usually, when I use a CLI tool, I'm told about each specific parameter that is missing/invalid. However, I'm not sure that's a better UX than this.

Reply:
I agree, that it might be superfluous to notify the user if they did not provide one out of three parameters about the other two which they did indeed provide. I also think the error message can be improved as it only references the `--` parameters but not their `-` variants.

Reply:
On the contrary I think this kind of message delivers the most concise requirement alas "you need to provide x y and z" - added the short names of the switches - what do you think?

Reply:
I like this.